src/org/mozilla/javascript/NativeConsole.java

```diff
@@ -0,0 +1,405 @@
+                break;
+            default:
+                throw new IllegalStateException(String.valueOf(id));
+        }
+        initPrototypeMethod(CONSOLE_TAG, id, name, arity);
+    }
+
+    @Override
+    public Object execIdCall(
+            IdFunctionObject f, Context cx, Scriptable scope, Scriptable thisObj, Object[] args) {
+        if (!f.hasTag(CONSOLE_TAG)) {
+            return super.execIdCall(f, cx, scope, thisObj, args);
+        }
+
+        int methodId = f.methodId();
+        switch (methodId) {
+            case Id_toSource:
+                return "Console";
+
+            case Id_trace:
+                printer.print(cx, scope, Level.TRACE, args);
+                break;
+
+            case Id_debug:
+                printer.print(cx, scope, Level.DEBUG, args);
+                break;
+
+            case Id_log:
+            case Id_info:
+                printer.print(cx, scope, Level.INFO, args);
+                break;
+
+            case Id_warn:
+                printer.print(cx, scope, Level.WARN, args);
+                break;
+
+            case Id_error:
+                printer.print(cx, scope, Level.ERROR, args);
+                break;
+
+            case Id_assert:
+                jsAssert(cx, scope, args);
+                break;
+
+            case Id_count:
+                count(cx, scope, args);
+                break;
+
+            case Id_countReset:
+                countReset(cx, scope, args);
+                break;
+
+            case Id_time:
+                time(cx, scope, args);
+                break;
+
+            case Id_timeEnd:
+                timeEnd(cx, scope, args);
+                break;
+
+            case Id_timeLog:
+                timeLog(cx, scope, args);
+                break;
+
+            default:
+                throw new IllegalStateException(String.valueOf(methodId));
+        }
+
+        return Undefined.instance;
+    }
+
+    private void print(Context cx, Scriptable scope, Level level, String msg) {
+        printer.print(cx, scope, level, new String[] {msg});
+    }
+
+    public static String format(Context cx, Scriptable scope, Object[] args) {
+        String msg = ScriptRuntime.toString(args[0]);
+        if (msg == null || msg.length() == 0) {
+            return "";
+        }
+
+        int argIndex = 1;
+        Matcher matcher = FMT_REG.matcher(msg);
+        StringBuffer buffer = new StringBuffer(msg.length() * 2);
+        while (matcher.find()) {
+            String placeHolder = matcher.group();
+            String replaceArg;
+            switch (placeHolder) {
+                case "%%":
+                    replaceArg = "%";
+                    break;
+
+                case "%s":
+                    replaceArg = ScriptRuntime.toString(args, argIndex);
+                    break;
+
+                case "%d":
+                case "%i":
+                    double number = ScriptRuntime.toNumber(args, argIndex);
```

Comment:
BigInt will cause an error.
```js
console.log('%d', 10n)
// js: uncaught JavaScript runtime exception: TypeError: Cannot convert BigInt to a number
```
The specifications are as follows.
https://console.spec.whatwg.org/#formatting-specifiers
> %d or %i | Element which substitutes is converted to an integer | %parseInt%(element, 10)

The implementation are as follows.

V8 (Node.js v14.17.0)
```js
console.log('%d', 10n) // 10n
```
SpiderMonkey (Firefox/97.0)
```js
console.log('%d', 10n) // undefined
```

Reply:
```js
console.log('%d', Symbol.iterator)
// js: uncaught JavaScript runtime exception: TypeError: The object is not a number
```
According to the specification, it seems that Symbol should be converted to NaN.
https://console.spec.whatwg.org/#formatter
> If Type(current) is Symbol, let converted be NaN

Reply:
Even `%o` or `%O` will cause an error.
```js
console.log('%o', { a: 0n }) 
// js: uncaught JavaScript runtime exception: TypeError: Do not know how to serialize a BigInt
```

Reply:
> ```js
> console.log('%d', 10n)
> // js: uncaught JavaScript runtime exception: TypeError: Cannot convert BigInt to a number
> ```

I got a syntax error. How to write a BigInt in rhino shell?

```
js> print(10n);
js: "<stdin>", line 10: missing ) after argument list
js: print(10n);
js: .........^
js: "<stdin>", line 10: Compilation produced 1 syntax errors.

js> console.log("%d", 10n);
js: "<stdin>", line 12: missing ) after argument list
js: console.log("%d", 10n);
js: .....................^
js: "<stdin>", line 12: Compilation produced 1 syntax errors.

js> 
```

Reply:
Could you add `-version 200`.
```sh
$ java -jar buildGradle/libs/rhino-1.7.15-SNAPSHOT.jar -version 200
```

Reply:
> Even `%o` or `%O` will cause an error.
> 
> ```js
> console.log('%o', { a: 0n }) 
> // js: uncaught JavaScript runtime exception: TypeError: Do not know how to serialize a BigInt
> ```

I think this is not the problem of console. This TypeError is thrown at `org.mozilla.javascript.NativeJSON#str`

```java
        if (value instanceof Number) {
            if (value instanceof BigInteger) {
                throw ScriptRuntime.typeErrorById("msg.json.cant.serialize", "BigInt");
            }
            // ...
        }
```
This problem also occurs when using JSON.stringify
```
js> JSON.stringify({a: 10n});
js: uncaught JavaScript runtime exception: TypeError: Do not know how to serialize a BigInt
	at (unknown):9
```

Reply:
Yes, I do. It is a specification of ECMA262 that `JSON.stringify` will error on BigInt.
https://262.ecma-international.org/12.0/#sec-serializejsonproperty
> 10. 10. If Type(value) is BigInt, throw a TypeError exception.

Even in Node.js, using `%j` will result in an error.
```js
console.log('%j', 10n)
// Uncaught TypeError: Do not know how to serialize a BigInt

console.log('%o', 10n) // 10n
```
Other than that, `JSON.stringify` will cause an error with circular references.
```js
var o = {};
o.o = o;
JSON.stringify(o);
// js: uncaught JavaScript runtime exception: TypeError: Cyclic org.mozilla.javascript.NativeObject value not allowed.
```

If you definitely don't want to generate errors, you should use a method other than `JSON.stringify` to display the data. However, for now, I think it is enough to catch the exception and fall back to `ScriptRuntime.toString`.